File: src/runtimeCapabilities.ts
export const RUNTIME_PLATFORMS = [
  "windows",
  "macos",
  "linux",
  "ios",
  "android",
  "other"
] as const;

export type RuntimePlatform = (typeof RUNTIME_PLATFORMS)[number];

export type RuntimeCapabilities = Readonly<{
  platform: RuntimePlatform;
  mobile: boolean;
  nativeFileDialogs: boolean;
  localPdfEditing: boolean;
  localVisualMarks: boolean;
  imageToPdf: boolean;
  externalProcesses: boolean;
  connectedScanning: boolean;
  cameraCapture: boolean;
  searchableOcr: boolean;
  certificateSigning: boolean;
  archivalPdf: boolean;
  passwordProtection: boolean;
  directUpdates: boolean;
  appStoreUpdates: boolean;
}>;

const booleanFields = [
  "mobile",
  "nativeFileDialogs",
  "localPdfEditing",
  "localVisualMarks",
  "imageToPdf",
  "externalProcesses",
  "connectedScanning",
  "cameraCapture",
  "searchableOcr",
  "certificateSigning",
  "archivalPdf",
  "passwordProtection",
  "directUpdates",
  "appStoreUpdates"
] as const satisfies readonly (keyof RuntimeCapabilities)[];
const capabilityFields = new Set<string>(["platform", ...booleanFields]);

export function parseRuntimeCapabilities(value: unknown): RuntimeCapabilities {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error("The runtime capability report is invalid.");
  }
  const record = value as Record<string, unknown>;
  if (
    Object.keys(record).length !== capabilityFields.size ||
    Object.keys(record).some((field) => !capabilityFields.has(field)) ||
    typeof record.platform !== "string" ||
    !RUNTIME_PLATFORMS.includes(record.platform as RuntimePlatform) ||
    booleanFields.some((field) => typeof record[field] !== "boolean")
  ) {
    throw new Error("The runtime capability report is invalid.");
  }

  const capabilities = record as RuntimeCapabilities;
  const expectedMobile = capabilities.platform === "ios" || capabilities.platform === "android";
  if (
    capabilities.mobile !== expectedMobile ||
    (capabilities.mobile && capabilities.externalProcesses) ||
    (capabilities.mobile && capabilities.connectedScanning) ||
    capabilities.appStoreUpdates !== (capabilities.platform === "ios") ||
    (capabilities.connectedScanning && !capabilities.externalProcesses) ||
    (capabilities.searchableOcr && !capabilities.externalProcesses) ||
    (capabilities.certificateSigning && !capabilities.externalProcesses) ||
    (capabilities.archivalPdf && !capabilities.externalProcesses) ||
    (capabilities.passwordProtection && !capabilities.externalProcesses)
  ) {
    throw new Error("The runtime capability report is inconsistent.");
  }
  return Object.freeze({ ...capabilities });
}

export function isAppleMobileRuntime(
  capabilities: RuntimeCapabilities | null
): boolean {
  return capabilities?.platform === "ios";
}

File: src-tauri/src/runtime_capabilities.rs
use serde::Serialize;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
#[serde(rename_all = "lowercase")]
pub(crate) enum RuntimePlatform {
    Windows,
    Macos,
    Linux,
    Ios,
    Android,
    Other,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct RuntimeCapabilities {
    platform: RuntimePlatform,
    mobile: bool,
    native_file_dialogs: bool,
    local_pdf_editing: bool,
    local_visual_marks: bool,
    image_to_pdf: bool,
    external_processes: bool,
    connected_scanning: bool,
    camera_capture: bool,
    searchable_ocr: bool,
    certificate_signing: bool,
    archival_pdf: bool,
    password_protection: bool,
    direct_updates: bool,
    app_store_updates: bool,
}

impl RuntimeCapabilities {
    pub(crate) fn external_processes(self) -> bool {
        self.external_processes
    }

    pub(crate) fn connected_scanning(self) -> bool {
        self.connected_scanning
    }

    pub(crate) fn searchable_ocr(self) -> bool {
        self.searchable_ocr
    }

    pub(crate) fn certificate_signing(self) -> bool {
        self.certificate_signing
    }

    pub(crate) fn archival_pdf(self) -> bool {
        self.archival_pdf
    }

    pub(crate) fn password_protection(self) -> bool {
        self.password_protection
    }
}

pub(crate) const fn capabilities_for(platform: RuntimePlatform) -> RuntimeCapabilities {
    let mobile = matches!(platform, RuntimePlatform::Ios | RuntimePlatform::Android);
    let desktop = matches!(
        platform,
        RuntimePlatform::Windows | RuntimePlatform::Macos | RuntimePlatform::Linux
    );
    RuntimeCapabilities {
        platform,
        mobile,
        native_file_dialogs: !matches!(platform, RuntimePlatform::Other),
        local_pdf_editing: !matches!(platform, RuntimePlatform::Other),
        local_visual_marks: !matches!(platform, RuntimePlatform::Other),
        image_to_pdf: !matches!(platform, RuntimePlatform::Other),
        external_processes: desktop,
        connected_scanning: desktop,
        camera_capture: false,
        searchable_ocr: desktop,
        certificate_signing: desktop,
        archival_pdf: desktop,
        password_protection: desktop,
        direct_updates: desktop,
        app_store_updates: matches!(platform, RuntimePlatform::Ios),
    }
}

pub(crate) const fn current_platform() -> RuntimePlatform {
    if cfg!(target_os = "windows") {
        RuntimePlatform::Windows
    } else if cfg!(target_os = "macos") {
        RuntimePlatform::Macos
    } else if cfg!(target_os = "linux") {
        RuntimePlatform::Linux
    } else if cfg!(target_os = "ios") {
        RuntimePlatform::Ios
    } else if cfg!(target_os = "android") {
        RuntimePlatform::Android
    } else {
        RuntimePlatform::Other
    }
}

pub(crate) const fn current_capabilities() -> RuntimeCapabilities {
    capabilities_for(current_platform())
}

#[tauri::command]
pub(crate) fn runtime_capabilities() -> RuntimeCapabilities {
    current_capabilities()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn desktop_targets_expose_subprocess_and_scanner_capabilities() {
        for platform in [
            RuntimePlatform::Windows,
            RuntimePlatform::Macos,
            RuntimePlatform::Linux,
        ] {
            let capabilities = capabilities_for(platform);
            assert!(!capabilities.mobile);
            assert!(capabilities.local_pdf_editing);
            assert!(capabilities.external_processes);
            assert!(capabilities.connected_scanning);
            assert!(capabilities.searchable_ocr);
            assert!(capabilities.certificate_signing);
            assert!(capabilities.direct_updates);
            assert!(!capabilities.app_store_updates);
        }
    }

    #[test]
    fn ios_keeps_the_local_pdf_core_without_desktop_process_claims() {
        let capabilities = capabilities_for(RuntimePlatform::Ios);
        assert!(capabilities.mobile);
        assert!(capabilities.native_file_dialogs);
        assert!(capabilities.local_pdf_editing);
        assert!(capabilities.local_visual_marks);
        assert!(capabilities.image_to_pdf);
        assert!(!capabilities.external_processes);
        assert!(!capabilities.connected_scanning);
        assert!(!capabilities.camera_capture);
        assert!(!capabilities.searchable_ocr);
        assert!(!capabilities.certificate_signing);
        assert!(!capabilities.archival_pdf);
        assert!(!capabilities.password_protection);
        assert!(!capabilities.direct_updates);
        assert!(capabilities.app_store_updates);
    }

    #[test]
    fn android_is_not_accidentally_reported_as_an_apple_store_build() {
        let capabilities = capabilities_for(RuntimePlatform::Android);
        assert!(capabilities.mobile);
        assert!(capabilities.local_pdf_editing);
        assert!(!capabilities.external_processes);
        assert!(!capabilities.connected_scanning);
        assert!(!capabilities.app_store_updates);
    }
}
